controllers/component_pipelinerun_nudge_controller.go

```diff
@@ -0,0 +1,220 @@
+		return ctrl.Result{}, nil
+	}
+
+	components := applicationapiv1alpha1.ComponentList{}
+	err = r.Client.List(ctx, &components, client.InNamespace(pipelineRun.Namespace))
+	if err != nil {
+		log.Error(err, "failed to list components in namespace")
+		return ctrl.Result{}, err
+	}
+
+	image := ""
+	digest := ""
+	for _, r := range pipelineRun.Status.PipelineResults {
+		if r.Name == ImageDigestParam {
+			digest = r.Value.StringVal
+		} else if r.Name == ImageUrlParam {
+			image = r.Value.StringVal
+		}
+	}
+	if image == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageUrlParam), "no image url result")
+		return ctrl.Result{}, nil
+	}
+	if digest == "" {
+		log.Error(fmt.Errorf("unable to find %s param on PipelineRun, not performing nudge", ImageDigestParam), "no image digest result")
+		return ctrl.Result{}, nil
+	}
+
+	for i := range components.Items {
+		comp := components.Items[i]
+		if comp.Spec.Application == updatedComponent.Spec.Application && slices.Contains(updatedComponent.Spec.BuildNudgesRef, comp.Name) {
+			//TODO: Do the nudge
+			log.Info(fmt.Sprintf("Nudging %s due to successful build of %s", comp.Name, updatedComponent.Name))
+			//TODO: do we want the event? It's just for unit testing at the moment
+			r.EventRecorder.Event(&comp, v12.EventTypeNormal, ComponentNudged, fmt.Sprintf("component %s.%s was nudged by successful build of %s that produces image %s@%s", comp.Namespace, comp.Name, updatedComponent.Name, image, digest))
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupController creates a new Integration reconciler and adds it to the Manager.
+func (r *ComponentNudgeReconciler) SetupWithManager(manager ctrl.Manager) error {
+	return setupControllerWithManager(manager, r)
+}
+
+// setupControllerWithManager sets up the controller with the Manager which monitors new PipelineRuns and filters
+// out status updates.
+func setupControllerWithManager(manager ctrl.Manager, reconciler *ComponentNudgeReconciler) error {
+
+	return ctrl.NewControllerManagedBy(manager).
+		For(&tektonapi.PipelineRun{}).
+		WithEventFilter(predicate.NewPredicateFuncs(IsBuildPushPipelineRun)).
+		Complete(reconciler)
+}
+
+func IsBuildPushPipelineRun(object client.Object) bool {
+	if pipelineRun, ok := object.(*tektonapi.PipelineRun); ok {
+		if pipelineRun.Labels != nil && pipelineRun.Annotations != nil {
+			if pipelineRun.Labels[PipelineRunTypeLabel] == PipelineRunBuildType && pipelineRun.Annotations[PacEventType] == PacEventPush {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// GetComponentFromPipelineRun loads from the cluster the Component referenced in the given PipelineRun. If the PipelineRun doesn't
+// specify a Component or this is not found in the cluster, an error will be returned.
+func GetComponentFromPipelineRun(c client.Client, ctx context.Context, pipelineRun *tektonapi.PipelineRun) (*applicationapiv1alpha1.Component, error) {
```

Comment:
Kinda, but we need the finalizer as well to prevent tekton results pruning the pipeline before we have reacted to it (if tekton results pruning ever gets enabled).

Reply:
How does the predicate influence the finalyzer?

Reply:
It will just need to pass through more events, as it also needs to see new runs.